crates/data_components/src/kafka.rs

```diff
@@ -132,13 +134,95 @@ impl std::fmt::Debug for KafkaConfig {
                 "ssl_endpoint_identification_algorithm",
                 &self.ssl_endpoint_identification_algorithm,
             )
+            .field(
+                "metrics_store",
+                &self.metrics_store.as_ref().map(|_| "Some(KafkaMetrics)"),
+            )
             .finish()
     }
 }
 
+#[derive(Debug, Default)]
+pub struct KafkaMetrics {
+    /// Total consumer lag across all partitions
+    pub records_lag: AtomicU64,
+    /// Total number of messages consumed
+    pub records_consumed: AtomicU64,
+    /// Total bytes consumed
+    pub bytes_consumed: AtomicU64,
+}
+
+struct KafkaConsumerContext {
+    metrics: Arc<KafkaMetrics>,
+}
+
+impl KafkaConsumerContext {
+    fn new(metrics: Arc<KafkaMetrics>) -> Self {
+        Self { metrics }
+    }
+}
+
+impl KafkaMetrics {
+    #[must_use]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn update_records_lag(&self, lag: u64) {
+        self.records_lag.store(lag, Ordering::Relaxed);
+    }
+
+    pub fn update_records_consumed(&self, count: u64) {
+        self.records_consumed.store(count, Ordering::Relaxed);
+    }
+
+    pub fn update_bytes_consumed(&self, bytes: u64) {
+        self.bytes_consumed.store(bytes, Ordering::Relaxed);
+    }
+}
+
+impl rdkafka::ClientContext for KafkaConsumerContext {
+    #[allow(clippy::cast_sign_loss)]
+    fn stats(&self, statistics: rdkafka::Statistics) {
+        // Calculate total consumer lag from all topic partitions
+        let mut total_lag = 0u64;
+        let mut has_valid_partitions = false;
+
+        for topic in statistics.topics.values() {
+            for partition in topic.partitions.values() {
+                // Skip internal partitions (partition id -1)
+                if partition.partition >= 0 {
+                    total_lag += partition.consumer_lag as u64;
+                    has_valid_partitions = true;
+                }
+            }
+        }
+
+        // Update total lag only if we have valid partitions to avoid misleading data
+        if has_valid_partitions {
+            self.metrics.update_records_lag(total_lag);
+        }
+
+        self.metrics
+            .update_records_consumed(statistics.rxmsgs as u64);
+        self.metrics
+            .update_bytes_consumed(statistics.rxmsg_bytes as u64);
+
+        tracing::trace!(
+            "Kafka metrics updated - lag: {}, messages: {}, bytes: {}",
```

Comment:
also include the topic(s) this is for

Reply:
Improved - thank you

>
2025-09-15T22:11:02.353850Z  TRACE data_components::kafka: Kafka metrics updated for consumer: rdkafka#consumer-1, topics: ["orders_events"], lag: 1, messages: 1544, bytes: 354932, brokers=["localhost:9092/1:UP", "GroupCoordinator:UP"], consumer_group_state=Some("up")

>
2025-09-15T22:15:07.348503Z  TRACE data_components::kafka: Kafka metrics updated for consumer: rdkafka#consumer-2, topics: ["orders_events2"], lag: 3, messages: 127, bytes: 29286, brokers=["localhost:9092/1:UP", "GroupCoordinator:UP"], consumer_group_state=Some("up")